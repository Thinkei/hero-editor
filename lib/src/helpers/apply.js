const apply = (args = []) => (func) => func(...args);

export default apply;
