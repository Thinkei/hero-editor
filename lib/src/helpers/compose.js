const compose = (...funcs) => (arg) =>
  funcs.reduceRight((composed, f) => (f ? f(composed) : composed), arg);

export default compose;
