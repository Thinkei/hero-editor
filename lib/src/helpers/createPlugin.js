const createPlugin = (configs) => {
  const defaultPlugin = {
    name: '',
    renderLeaf: (next) => (props) => next(props),
    renderElement: (next) => (props) => next(props),
    ToolbarButton: () => null,
    renderCustom: () => null,
    handleMessage: () => () => {},
    enhanceEditor: (editor) => editor,
  };
  return { ...defaultPlugin, ...configs };
};

export default createPlugin;
