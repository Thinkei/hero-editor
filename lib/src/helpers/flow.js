const flow = (...funcs) => (arg) =>
  funcs.reduce((composed, f) => (f ? f(composed) : composed), arg);

export default flow;
