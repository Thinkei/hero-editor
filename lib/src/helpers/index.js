import React from 'react';
import { Editor } from 'slate';
import { NUMBERED_LIST, BULLETED_LIST } from '../constants';

const createPlugin = (configs) => {
  const defaultPlugin = {
    name: '',
    renderLeaf: (next) => (props) => next(props),
    renderElement: (next) => (props) => next(props),
    handleMessage: () => () => {},
    ToolbarButton: () => null,
    renderCustom: () => null,
  };
  return { ...defaultPlugin, ...configs };
};

const isMarkActive = (editor, format) => {
  const marks = Editor.marks(editor);
  return marks ? marks[format] === true : false;
};

const toggleMark = (editor, format) => {
  const isActive = isMarkActive(editor, format);

  if (isActive) {
    Editor.removeMark(editor, format);
  } else {
    Editor.addMark(editor, format, true);
  }
};

const makeRenderLeaf = (mark, CustomLeaf) => (next) => (props) => {
  if (props.leaf[mark]) {
    const children = <CustomLeaf {...props} />;
    return next({ ...props, children });
  }

  return next(props);
};

const DefaultLeaf = 'span';

const renderDefaultLeaf = ({ attributes, children }) => (
  <DefaultLeaf {...attributes}>{children}</DefaultLeaf>
);

const compose = (...funcs) => (arg) =>
  funcs.reduceRight((composed, f) => (f ? f(composed) : composed), arg);

const composeRenderLeaf = (...renderers) =>
  compose(...renderers)(renderDefaultLeaf);

const messagePrefix = '@hero-editor';

const addMessageListener = (type, callback) => (editor) => {
  const listener = (event) => {
    if (event.data.type === `${messagePrefix}/${editor.id}/${type}`) {
      callback({ editor, data: event.data.data });
    }
  };
  window.addEventListener('message', listener);
  return () => window.removeEventListener('message', listener);
};

const postMessage = (type, data, editor) =>
  window.postMessage({
    type: `${messagePrefix}/${editor.id}/${type}`,
    data,
  });

const DefaultElement = React.forwardRef((props, ref) => (
  <p ref={ref} style={{ margin: 0, lineHeight: 1.5 }} {...props} />
));

const renderDefaultElement = ({ attributes, children }) => (
  <DefaultElement {...attributes}>{children}</DefaultElement>
);

const composeRenderElement = (...renderers) =>
  compose(...renderers)(renderDefaultElement);

const makeRenderElement = (block, CustomElement) => (next) => (props) => {
  if (props.element.type === block) {
    return (
      <CustomElement {...props.attributes}>{props.children}</CustomElement>
    );
  }

  return next(props);
};

const isBlockActive = (editor, format) => {
  const [match] = Editor.nodes(editor, {
    match: (n) => n.type === format,
  });

  return !!match;
};

const isList = (node) => [BULLETED_LIST, NUMBERED_LIST].includes(node.type);

export {
  createPlugin,
  isMarkActive,
  toggleMark,
  makeRenderLeaf,
  renderDefaultLeaf,
  compose,
  composeRenderLeaf,
  addMessageListener,
  postMessage,
  composeRenderElement,
  renderDefaultElement,
  makeRenderElement,
  isBlockActive,
  isList,
};
