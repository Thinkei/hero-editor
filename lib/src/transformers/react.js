import React from 'react';
import { Text, Node } from 'slate';
import y from '../helpers/why';

const isRoot = Array.isArray;

const makeReactTransformer = (rules) =>
  y((reactTransformer) => (node, key) => {
    if (isRoot(node)) return node.map(reactTransformer);

    if (Text.isText(node)) return rules.leaf(node, key);

    const children = node.children.map(reactTransformer);

    return rules.element(node, children, key);
  });

const defaultRules = {
  leaf: (node, key) => {
    let children = node.text;

    if (node.bold) children = <strong key={key}>{children}</strong>;

    if (node.italic) children = <em key={key}>{children}</em>;

    if (node.underline) children = <u key={key}>{children}</u>;

    return <span key={key}>{children}</span>;
  },

  element: (node, children, key) => {
    switch (node.type) {
      case 'paragraph':
        return (
          <p
            key={key}
            style={{ margin: 0, lineHeight: 1.5, minHeight: '1.5em' }}
          >
            {children}
          </p>
        );

      case 'heading-one':
        return <h1 key={key}>{children}</h1>;

      case 'link':
        return (
          <a key={key} href={Node.string(node)} target="_blank">
            {children}
          </a>
        );

      case 'mention':
        return (
          <span
            key={key}
            style={{
              padding: '2px 4px',
              background: '#eeeeee',
              borderRadius: 4,
            }}
          >
            @{node.data.name}
          </span>
        );

      case 'bulleted-list':
        return <ul key={key}>{children}</ul>;

      case 'numbered-list':
        return <ol key={key}>{children}</ol>;

      case 'list-item':
        return <li key={key}>{children}</li>;

      default:
        return <div key={key}>{children}</div>;
    }
  },
};

const defaultReactTransformer = makeReactTransformer(defaultRules);

export default makeReactTransformer;

export { makeReactTransformer, defaultReactTransformer };
