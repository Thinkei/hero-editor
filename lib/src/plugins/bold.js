import React from 'react';
import { useSlate } from 'slate-react';
import { isMarkActive, toggleMark, makeRenderLeaf } from '../helpers';

const renderLeaf = makeRenderLeaf('bold', ({ children }) => (
  <strong>{children}</strong>
));

const ToolbarButton = () => {
  const editor = useSlate();

  return (
    <button
      style={{
        background: isMarkActive(editor, 'bold') ? 'grey' : 'lightgrey',
      }}
      onMouseDown={(event) => {
        event.preventDefault();
        toggleMark(editor, 'bold');
      }}
    >
      bold
    </button>
  );
};

export default { renderLeaf, ToolbarButton };
