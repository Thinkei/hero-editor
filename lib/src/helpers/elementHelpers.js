import React from 'react';
import { Editor } from 'slate';
import compose from './compose';
import { NUMBERED_LIST, BULLETED_LIST, HEADING } from '../constants';

const isBlockActive = (editor, format) => {
  const [match] = Editor.nodes(editor, {
    match: (n) => n.type === format,
  });
  return !!match;
};

const isList = (node) => [BULLETED_LIST, NUMBERED_LIST].includes(node.type);

const DefaultElement = React.forwardRef((props, ref) => (
  <p ref={ref} style={{ margin: 0, lineHeight: 1.5 }} {...props} />
));

const HeadingElement = React.forwardRef((props, ref) => (
  <div ref={ref} style={{ margin: 0, lineHeight: 1.5 }} {...props} />
));

const renderDefaultElement = ({ attributes, children, element }) => {
  const isHeading = element.type === HEADING;
  return isHeading ? (
    <HeadingElement {...attributes}>{children}</HeadingElement>
  ) : (
    <DefaultElement {...attributes}>{children}</DefaultElement>
  );
};

const makeRenderElement = (block, CustomElement) => (next) => (props) => {
  if (props.element.type === block) {
    return <CustomElement {...props} />;
  }
  return next(props);
};

const composeRenderElement = (plugins) =>
  compose(...plugins.map((plugin) => plugin.renderElement))(
    renderDefaultElement,
  );

export {
  isBlockActive,
  isList,
  renderDefaultElement,
  makeRenderElement,
  composeRenderElement,
};
