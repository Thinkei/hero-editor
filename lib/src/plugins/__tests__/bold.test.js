import { render } from '@testing-library/react';
import bold from '../bold';
import { renderDefaultLeaf } from '../../helpers';

describe('plugin bold', () => {
  it('is a plugin object', () => {
    expect(typeof bold.renderLeaf).toBe('function');
  });

  it('renders bold element', () => {
    const { baseElement } = render(
      bold.renderLeaf(renderDefaultLeaf)({
        leaf: { bold: true },
        children: 'Bold text',
      }),
    );
    expect(baseElement).toMatchSnapshot();
  });
});
