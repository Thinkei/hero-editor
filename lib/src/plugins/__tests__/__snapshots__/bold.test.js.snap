// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin bold renders bold element 1`] = `
<body>
  <div>
    <span>
      <strong>
        Bold text
      </strong>
    </span>
  </div>
</body>
`;
